### PR TITLE
Add presets for recording traces for crbug issues

### PR DESCRIPTION
**Pull request by kphanee**
WIP: Not for landing. Posted for further discussion.

### CODE REVIEW

**Comment by kphanee** [comment]
@natduca , @dj2 : 
TLDR: Please go through the entire comment. :). An initial WIP dump for issue #684 - adding 2 new preset modes for crbug.

My original understanding of the requirement is:
*All default on categories* implies **All categories excluding all (disabled-by-default- and test categories)**
*All default on categories plus disabled-by-default-cc.debug* implies **All categories including disabled-by-default-cc.debug and excluding all (other disabled-by-default-) and test categories**
Correct me if I am wrong in my interpretation. 

As per the native code, the default category filter is: "-*Debug, -*Test".
On checking the default scenario for android (using script), the default string is passed "as it is" to TraceLog class (category_filter_). Should I do the same from js as well? Is that is what we are expecting?

I have also kept another approach (commented) where I am excluding all the categories which either have 'debug' or 'test' present in the category string. Is this the expected approach? 

I also wanted to understand the current logic of the filling the 'categories' array.
```javascript
        for (var category in allCategories) {
          var disabled = category.indexOf('disabled-by-default-') == 0;
          if (this.currentlyChosenPreset_.indexOf(category) >= 0) {
            if (disabled)
              categories.push(category);
          } else {
            if (!disabled)
              categories.push('-' + category);
          }
        }
```
As I understand, this is the flow:
* For each category present in allCategories object:
  * Check if it is a disabled by default category or not.
  * Is category present in the preset array?
    * If yes, if it is disabled, add it to final array. **Why adding only disabled-by-default category?**
    * If no, is it disabled by default category?
      * If no, then add it final array as excluding category. **I could understand this action**

As per my understanding, the way TraceLog::category_filter_ works is:
* All required categories should be in inclusion syntax
* All other categories should be in exclusion syntax (preceded by '-').
All the regular categories are by default included and all disabled-by-default categories are by default excluded. Please correct me if I am wrong. Hence, the logic that I copied above. So, should I fill the array accordingly even in case of regex expression presence?

I was trying to debug into TraceLog class to understand how the category_filter_ is being used, but couldn't head much. It would be great if either of you can give me some insight how the category_filter_ is being used.

**Comment by natduca** [comment]
The result of RecordingSelectionDialog() is selectionDlg.categoryFilter(), which is a string.

That string is a comma selected list of categories to include and exclude.


Preset 1: All default on categories. Dialog should return çateogryFilter() = ''
Preset 2: All default on categorires plus cc-debug: Dialog should return categoryFilter() = 'disabled-by-default-cc.debug'

Your job is to refactor the recording selection dialog implementation so that this happens.

**Comment by natduca** [comment]
Oh, maybe this'll help:
http://src.chromium.org/chrome/trunk/src/tools/telemetry/telemetry/core/platform/tracing_category_filter.py

Maybe the first thing to do to do an implementation of this class in javascript, and modify RSD and record_and_capture_controller to use that object. That way the logic of the category filter is defined in a single class instead of mixed in with UI code.

Then, as a followup, you can then have presets that just provide categoryFilters...

Dunno, maybe thats not the right direction. Thought I'd throw it out there.

**Comment by dj2** [comment]
Is this closed or still in progress?

**Comment by kphanee** [comment]
This CL is still open and needs to be updated.
I didn't work on this in recent times on the lines @natduca suggested.
I will update this CL with new changes.

**Comment by dj2** [comment]
Closing this as it's been several months. Please send a CL up to rietveld if this should still be open.
